fpdf/text_region.py

```diff
@@ -0,0 +1,428 @@
+import math
+from typing import NamedTuple, Sequence
+
+from .errors import FPDFException
+from .enums import Align, XPos, YPos, WrapMode
+from .line_break import MultiLineBreak
+
+# Since Python doesn't have "friend classes"...
+# pylint: disable=protected-access
+
+
+class Extents(NamedTuple):
+    left: float
+    right: float
+
+
+class TextRegionMixin:
+    """Mix-in to be added FPDF() in order to support text regions."""
+
+    def __init__(self, *args, **kwargs):
+        self.clear_text_region()
+        super().__init__(*args, **kwargs)
+
+    def register_text_region(self, region):
+        self.__current_text_region = region
+
+    def is_current_text_region(self, region):
+        return self.__current_text_region == region
+
+    def clear_text_region(self):
+        self.__current_text_region = None
+
+
+class Paragraph:
+    def __init__(
+        self,
+        region,
+        align=None,
+        line_height=None,
+        top_margin: float = 0,
+        bottom_margin: float = 0,
+        skip_leading_spaces: bool = False,
+        wrapmode: WrapMode = None,
+    ):
+        self._region = region
+        self.pdf = region.pdf
+        if align:
+            align = Align.coerce(align)
+        self.align = align
+        if line_height is None:
+            self.line_height = region.line_height
+        else:
+            self.line_height = line_height
+        self.top_margin = top_margin
+        self.bottom_margin = bottom_margin
+        self.skip_leading_spaces = skip_leading_spaces
+        if wrapmode is None:
+            self.wrapmode = self._region.wrapmode
+        else:
+            self.wrapmode = WrapMode.coerce(wrapmode)
+        self._text_fragments = []
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._region.end_paragraph()
+
+    def write(self, text: str, link=None):
+        if not self.pdf.font_family:
+            raise FPDFException("No font set, you need to call set_font() beforehand")
+        normalized_string = self.pdf.normalize_text(text).replace("\r", "")
+        # YYY _preload_font_styles() should accept a "link" argument.
+        fragments = self.pdf._preload_font_styles(normalized_string, False)
+        if link:
+            for frag in fragments:
+                frag.link = link
+        self._text_fragments.extend(fragments)
+
+    def ln(self, h=None):
+        if not self.pdf.font_family:
+            raise FPDFException("No font set, you need to call set_font() beforehand")
+        if h is None:
+            h = self.pdf.font_size * self.line_height
+        fragment = self.pdf._preload_font_styles("\n", False)[0]
+        fragment.graphics_state["font_size_pt"] = h * fragment.k
+        self._text_fragments.append(fragment)
+
+    def build_lines(self, print_sh):
```

Comment:
```suggestion
    def build_lines(self, print_sh) -> List[LineWrapper]:
```
Documenting the return type there would slightly improve the code readability